src/vmm/src/devices/virtio/device.rs

```diff
@@ -36,7 +36,7 @@ impl DeviceState {
     }
 
     /// Gets the memory and interrupt attached to the device if it is activated.
-    pub fn active_state(&self) -> Option<(&GuestMemoryMmap, Arc<IrqTrigger>)> {
+    pub fn active_state(&self) -> Option<(&GuestMemoryMmap, Arc<dyn VirtioInterrupt>)> {
```

Comment:
I see why you use dynamic dispatch here, and maybe it is fine from a perf perspective, but I wonder if we can somehow avoid it. Technically any device will be either `MMIO` or `PCIe`. So maybe we can make them generic over `VirtioInterrupt`? This will be a bigger change, but should not be a difficult one. Eventually all devices are abstracted as `VirtioDevice` anyway.

Reply:
Yup, that's not a bad idea. However, all this re-working is not going to go away. We will need to pass the `VirtioInterrupt` during activation (that's how PCIe devices do it).

Reply:
I am fine with passing it during activation. I am mainly concerned with dynamic dispatch. If we were to pass `&impl VirtioInterrupt` it would be better I think. In any case, this can be discussed later, when first PR with PCIe has arrived.

Reply:
> I don't think we necessarily can, because then we'd have to decide at compile time whether we use MMIO or PCIe, right?

But I think we have two different execution paths. One through a MMIO device manager and one through a PCIe device manager. So both variants would be compiled in. Am I getting this wrong?